addon/PageScraper.js

```diff
@@ -30,33 +38,63 @@ function PageScraper(previewProvider, tabTracker, options = {}) {
 
 PageScraper.prototype = {
   /**
-   * Check if the link is already in the metadata database. If not, parse the
-   * HTML and insert it in the metadata database
+   * Parse the HTML and attempt to insert it in the metadata database
    */
-  _asyncParseAndSave: Task.async(function*(rawHTML, url, event) {
-    let startTime = Date.now();
-    this._tabTracker.handlePerformanceEvent(event, "metadataReceivedRawHTML", startTime);
-
-    let link = yield this._previewProvider.asyncLinkExist(url);
-    if (!link) {
-      const metadata = this._metadataParser.parseHTMLText(rawHTML, url);
-      if (this._shouldSaveMetadata(metadata)) {
-        try {
-          metadata.images = yield this._computeImageSize(metadata);
-        } catch (e) {
-          Cu.reportError(`PageScraper failed to compute image size for ${url}`);
-        }
-        this._insertMetadata(metadata);
-        this._tabTracker.handlePerformanceEvent(event, "metadataParseSuccess", Date.now() - startTime);
-      } else {
-        this._tabTracker.handlePerformanceEvent(event, "metadataInvalidReceived", Date.now() - startTime);
+  _asyncParseAndSave(rawHTML, url, event) {
+    this._tabTracker.handlePerformanceEvent(event, PERFORMANCE_EVENT_NAMES.metadata_raw_html, Date.now());
+    const metadata = this._metadataParser.parseHTMLText(rawHTML, url);
+    this._asyncSaveMetadata(metadata, event);
+  },
+
+  /**
+   * Save the metadata in the MetadataStore DB
+   */
+  _asyncSaveMetadata: Task.async(function*(metadata, event) {
+    const startTime = Date.now();
+    let linkExists = yield this._previewProvider.asyncLinkExist(metadata.url);
+    if (linkExists) {
+      this._tabTracker.handlePerformanceEvent(event, PERFORMANCE_EVENT_NAMES.metadata_exists, Date.now());
+      return;
+    }
+    if (!this._shouldSaveMetadata(metadata)) {
+      this._tabTracker.handlePerformanceEvent(event, PERFORMANCE_EVENT_NAMES.metadata_invalid, Date.now() - startTime);
+      return;
+    }
+    try {
+      metadata.images = yield this._computeImageSize(metadata);
+    } catch (e) {
+      Cu.reportError(`PageScraper failed to compute image size for ${metadata.url}`);
+    }
+    this._insertMetadata(metadata);
```

Comment:
you mean just call previewprovider's method directly here instead of adding an extra function call?